libavfilter/vf_overlay_videotoolbox.m

```diff
@@ -243,13 +262,14 @@ static int overlay_vt_blend(FFFrameSync *fs) API_AVAILABLE(macos(10.11), ios(9.0
 static av_cold void do_uninit(AVFilterContext *avctx) API_AVAILABLE(macos(10.11), ios(9.0))
 {
     OverlayVideoToolboxContext *ctx = avctx->priv;
+
+    ff_framesync_uninit(&ctx->fs);
+    av_buffer_unref(&ctx->device_ref);
```

Comment:
Actually we have to keep something like `output_configured` because `do_uninit` may be called even before the output is successfully configured.

Reply:
Without check if this actually configured, this will free an address not allocated. The `if(ctx->device_ref)` won't work because this address is not null, but not allocated yet.

Reply:
I don't think adding an extra reference to (ctx->device_ref) is even necessary.
[av_hwframe_ctx_alloc](https://github.com/FFmpeg/FFmpeg/blob/58ffe0db4d50a5aa633d4efb9b53c879aedcd1e1/libavutil/hwcontext.c#L279) + [hwframe_ctx_free](https://github.com/FFmpeg/FFmpeg/blob/58ffe0db4d50a5aa633d4efb9b53c879aedcd1e1/libavutil/hwcontext.c#L244) already handle this internally.

Reply:
That looks good. Removed the reference in struct.

Reply:
IMHO these lines can be dropped. https://github.com/gnattu/jellyfin-ffmpeg/blob/vf_overlay_videotoolbox/libavfilter/vf_overlay_videotoolbox.m#L433-L437

The buf of `link->hw_frames_ctx` hold an reference to the `frames_ctx->device_ref` of inlink buf.

See also:
- https://github.com/FFmpeg/FFmpeg/blob/cffadfe99a1c41866701bab6c7410c17737a1db1/libavfilter/qsvvpp.c#L581
- https://github.com/FFmpeg/FFmpeg/blob/cffadfe99a1c41866701bab6c7410c17737a1db1/libavfilter/qsvvpp.c#L608

Reply:
With more extensive testing, it appears that neither the `device_ref` in the struct nor the `av_buffer_ref` call can be removed for some video over video input combinations. If the `device_ref` is only managed by `hwframe_ctx` itself, we encounter outright segmentation faults for certain video inputs. It seems that the context is dropped prematurely while it is still in use.

And if we wish to manage the `device_ref` ourselves, we will need to use `av_buffer_ref` to obtain a new reference and release it at our convenience.

It is actually this line we can just borrow the reference and no need to use `av_buffer_ref `:
https://github.com/gnattu/jellyfin-ffmpeg/blob/de3f3d3ebfc9beda49965300259c186504d6ce21/libavfilter/vf_overlay_videotoolbox.m#L429

Reply:
You deal with it, as long as it satisfies your various test cases. Since I don't have a test environment I can't comment much on this.

Reply:
Now I believe we need to default to at least `MTLPixelFormatRGBA16Float` and `kCVPixelFormatType_64RGBAHalf` for all inputs, not only 10-bit inputs. The current `kCVPixelFormatType_32BGRA` does not cover all colors in even 8-bit YUV planes, resulting in slightly off colors in the final video. This is noticeable in some videos, especially when `repeatlast=0`, as there will be no color conversion after the overlay disappears, leading to a easily noticeable color change at that moment. Using full-range `kCVPixelFormatType_64RGBALE`, which supports 65535 samples, or using half-precision float sampled `kCVPixelFormatType_64RGBAHalf`, both mitigate the color problem.

Reply:
Interesting issue. If I understand correctly, there are inaccurate colors in the blending path, but the correct colors are in the copying path.

Is it also possible that it is a [limited vs full](https://github.com/gnattu/jellyfin-ffmpeg/blob/0c3363bc5ced646ed4fca339105fe2223b160468/libavutil/hwcontext_videotoolbox.c#L42) range issue? resulting in inaccurate yuv2rgb conversion matrix parameters inside the VT/CV transfer functions.

What's the performance penalty if we go this route?

Reply:
> What's the performance penalty if we go this route?
No measurable performance penalty on M1 Max, will use more memory for sure but it's still using less than what a browser would use so I'm fine with that.

> Is it also possible that it is a [limited vs full](https://github.com/gnattu/jellyfin-ffmpeg/blob/0c3363bc5ced646ed4fca339105fe2223b160468/libavutil/hwcontext_videotoolbox.c#L42) range issue?

Not quite, as `kCVPixelFormatType_32BGRA` is already a full-range color format. I think it is more likely to be a "sample rate" issue. The 8-bit sample cannot accurately represent all colors that the 8-bit YUV could. This means that some colors in the 8-bit YUV space simply do not exist in the 8-bit RGB space(and vice versa), and if this occurs, our conversion is no longer lossless as a different but close color has to be picked instead. The double conversion in our pipeline (YUV -> RGB -> YUV) makes it even more noticeable.

This is a bad example(in hevc, sorry if it cannot be played in your browser):

https://github.com/gnattu/jellyfin-ffmpeg/assets/12995396/fc4e4e24-94e4-4bc9-8b04-486789acd977

And this is the good example:

https://github.com/gnattu/jellyfin-ffmpeg/assets/12995396/cc98685c-a469-44d1-ada0-af4275df8ded

Notice the bad example looks darker when the overlay is playing, but it turns brighter when the overlay disappears.

Reply:
Chrome has supported HEVC since 2022, and Firefox did the same thing not long ago, but it is all through hardware acceleration and software decoding is not available.

You are probably right, the conversion of YUV to RGB is irreversible in some cases. So the brighter version looks almost similar to the source film, right?

It is indeed not a problem for beasts like the M1 Max. But for legacy Intel HD/UHD graphics card models (usually < ~0.5 TFLOPS FP32), when multiple HW filters are used in parallel, including tone-mapping filters, there may be a visible FPS loss.

Reply:
> So the brighter version looks almost similar to the source film, right?

Yes it is. The frames after the overlay is gone are not color-converted but only copied. As you can see in the good example, it is not distinguishable by human eyes.

> But for legacy Intel HD/UHD graphics card models (usually < ~0.5 TFLOPS FP32), when multiple HW filters are used in parallel, including tone-mapping filters, there may be a visible FPS loss.

It could be, but I *think* a precise color is more important.

Reply:
Then all is done. Try to submit it to [ffmpeg-devel](https://ffmpeg.org/pipermail/ffmpeg-devel/) and cc [quink-black](https://github.com/quink-black) with [his email](https://github.com/FFmpeg/FFmpeg/blob/37936b09ce316c32c456539afeb748d472088135/.mailmap#L5). If you want to test it for a while longer, you can also complete the remaining HW subtitle burn-in logic in Jellyfin.

Btw Jellyfin team currently lacks developers to maintain our FFmpeg fork on MacOS, including Github CI and packaging. Because members using MacOS don't know much about FFmpeg and video streaming, the rest are on Linux. Our fork mainly includes patches to solve some long-standing issues in upstream, and some custom modifications that are not suitable for upstream. In the future, we will also rewrite the entire transcoder of Jellyfin, and we may also need help from MacOS developers.

If you are interested in this, I can ask our admin to invite you to join. It doesn't matter if you don't have time. Feel free to contribute random pull requests. Matrix rooms is preferred to contact us https://jellyfin.org/contact

Reply:
I already sent it: https://patchwork.ffmpeg.org/project/ffmpeg/list/?series=10848
I forgot to cc that guy though.

> If you are interested in this, I can ask our admin to invite you to join.

I'm very happy to help. I personally running Jellyfin on Mac as my media server so getting things fixed can also benefits myself.

Reply:
> Matrix rooms is preferred to contact us

Just joined with `@gnattu:matrix.org`

Reply:
It should be this one:

https://patchwork.ffmpeg.org/project/ffmpeg/list/?series=10850

The v1 patch includes unresolved merge conflict.

Reply:
It seems that only part of the patchset is sent successfully. Also I forgot to mention, it's better to separate utils/filter/docs into different patches.

Reply:
> I already sent it: https://patchwork.ffmpeg.org/project/ffmpeg/list/?series=10848 I forgot to cc that guy though.
> 
> > If you are interested in this, I can ask our admin to invite you to join.
> 
> I'm very happy to help. I personally running Jellyfin on Mac as my media server so getting things fixed can also benefits myself.

Great! Due to the time difference, the admins are not always online. I'll let you know when I get an update.

Reply:
> It seems that only part of the patchset is sent successfully. Also I forgot to mention, it's better to separate utils/filter/docs into different patches.

v3 is incoming: https://patchwork.ffmpeg.org/project/ffmpeg/list/?series=10851

Reply:
All that remains is to keep an eye on it. If you still don't get a review after a few days, you may need to manually ping the reviewer.

In addition, our admin should have contacted you not long ago.